### PR TITLE
[NOD-773] postgres + go-pg migration

### PR DESCRIPTION
**Pull request by dereeno**

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -11,15 +11,19 @@ import (
...
 
 // KasparovFlags holds configuration common to both the Kasparov server and the Kasparov daemon.
 type KasparovFlags struct {
 	ShowVersion bool   `short:"V" long:"version" description:"Display version information and exit"`
 	LogDir      string `long:"logdir" description:"Directory to log output."`
 	DebugLevel  string `short:"d" long:"debuglevel" description:"Set log level {trace, debug, info, warn, error, critical}"`
-	DBAddress   string `long:"dbaddress" description:"Database address"`
+	DBHost      string `long:"dbhost" description:"Database host"`
+	DBPort      string `long:"dbport" description:"Database port"`
+	DBSSLMode   string `long:"dbsslmode" description:"Database SSL mode {disable, allow, prefer, require, verify-ca, verify-full}"`
```
Comment (someone235) [inline config/config.go]: This field is not used anywhere

File: database/database.go [hunk 2]
```diff
@@ -52,33 +41,25 @@ func Connect(cfg *config.KasparovFlags) error {
 		return errors.Errorf("Database is not current (version %d). Please migrate"+
 			" the database by running the server with --migrate flag and then run it again", version)
 	}
-
-	db, err = gorm.Open("mysql", connectionString)
-	if err != nil {
-		return err
-	}
-	db.SetLogger(gormLogger{})
+	db = pg.Connect(&pg.Options{
```
Comment (someone235) [inline database/database.go]: Why cfg.DBHost is ignored here?

File: database/migrations/000001_create_blocks_table.up.sql [hunk 3]
```diff
@@ -1,23 +1,24 @@
...
\ No newline at end of file
+    id                      BIGSERIAL,
+    block_hash              CHAR(64) NOT NULL,
+    accepting_block_id      BIGINT CHECK (accepting_block_id > 0) NULL,
+    version                 INT NOT NULL,
+    hash_merkle_root        CHAR(64) NOT NULL,
+    accepted_id_merkle_root CHAR(64) NOT NULL,
+    utxo_commitment         CHAR(64) NOT NULL,
+    timestamp               TIMESTAMP(0) NOT NULL,
+    bits                    BIGINT CHECK (bits >= 0) NOT NULL,
```
Comment (someone235) [inline database/migrations/000001_create_blocks_table.up.sql]: Can you also add upper bound of maxuint32?

File: dbaccess/block.go [hunk 4]
```diff
@@ -15,71 +15,64 @@ func BlockByHash(ctx Context, blockHash string, preloadedFields ...dbmodels.Fiel
...
+		return nil, err
 	}
 
 	return blocks, nil
 }
 
 // Blocks retrieves from the database up to `limit` blocks in the requested `order`, skipping the first `skip` blocks
 // If preloadedFields was provided - preloads the requested fields
-func Blocks(ctx Context, order Order, skip uint64, limit uint64, preloadedFields ...dbmodels.FieldName) ([]*dbmodels.Block, error) {
+func Blocks(ctx Context, order Order, skip int, limit int, preloadedFields ...dbmodels.FieldName) ([]*dbmodels.Block, error) {
```
Comment (someone235) [inline dbaccess/block.go]: I'm not sure about it. Logically the API should expose uint, and then internally it should be converted to whatever needed by go-pg. We can argue about it if you want.

File: dbaccess/block.go [hunk 5]
```diff
@@ -15,71 +15,64 @@ func BlockByHash(ctx Context, blockHash string, preloadedFields ...dbmodels.Fiel
...
 
-	query := db.
+	blocks := []*dbmodels.Block{}
+	query := db.Model(&blocks).
 		Offset(skip).
 		Limit(limit)
 
 	if order != OrderUnknown {
-		query = query.Order(fmt.Sprintf("`id` %s", order))
+		query = query.Order(fmt.Sprintf("block.id %s", order))
```
Comment (someone235) [inline dbaccess/block.go]: Why did you change to block.id?

Comment (someone235) [reply dbaccess/block.go]: And we should consider keeping constants for SQL field names

File: dbaccess/block.go [hunk 6]
```diff
@@ -93,16 +86,16 @@ func SelectedTip(ctx Context) (*dbmodels.Block, error) {
...
-		Where(&dbmodels.Block{IsChainBlock: true}).
-		First(block)
+	err = db.Model(block).
+		Where("is_chain_block = ?", true).
+		Order("blue_score DESC").
+		First()
 
-	dbErrors := dbResult.GetErrors()
-	if httpserverutils.IsDBRecordNotFoundError(dbErrors) {
+	if err == pg.ErrNoRows {
```
Comment (someone235) [inline dbaccess/block.go:96]: Please delete IsDBRecordNotFoundError and related functions

File: dbaccess/block.go [hunk 7]
```diff
@@ -115,21 +108,16 @@ func SelectedTipBlueScore(ctx Context) (uint64, error) {
...
-		Pluck("blue_score", &blueScore)
-
-	dbErrors := dbResult.GetErrors()
-	if httpserverutils.HasDBError(dbErrors) {
-		return 0, httpserverutils.NewErrorFromDBErrors("some errors were encountered when loading selected tip blue score from the database:", dbErrors)
+	var blueScore uint64
+	err = db.Model(&dbmodels.Block{}).
+		Where("is_chain_Block = ?", true).
+		ColumnExpr("MAX(blue_score) as blue_score").
+		Select(&blueScore)
```
Comment (someone235) [inline dbaccess/block.go:117]: Shouldn't you use .First or something like this?

File: dbaccess/common.go [hunk 8]
```diff
@@ -25,10 +23,9 @@ func Save(ctx Context, value interface{}) error {
 		return err
 	}
 
-	dbResult := db.Save(value)
-	dbErrors := dbResult.GetErrors()
-	if httpserverutils.HasDBError(dbErrors) {
-		return httpserverutils.NewErrorFromDBErrors("failed to save object: ", dbErrors)
+	err = db.Insert(value)
```
Comment (someone235) [inline dbaccess/common.go]: The comment says that `Save` updates, but here you insert. Please update the comment and check that changing it to `Insert`  doesn't harm anything.

File: dbaccess/transaction.go [hunk 9]
```diff
@@ -69,23 +64,22 @@ func TransactionsByAddress(ctx Context, address string, order Order, skip uint64
...
+	query := db.Model(&txs)
+	query = joinTxInputsTxOutputsAndAddresses(query).
+		Where("out_addresses.address = ?", address).
+		WhereOr("in_addresses.address = ?", address).
+		Limit(int(limit)).
+		Offset(int(skip))
+
 	if order != OrderUnknown {
-		query = query.Order(fmt.Sprintf("`transactions`.`id` %s", order))
+		query = query.Order(fmt.Sprintf("id %s", order))
```
Comment (someone235) [inline dbaccess/transaction.go]: isn't `id` ambiguous with join?

File: dbaccess/transaction.go [hunk 10]
```diff
@@ -98,40 +92,41 @@ func TransactionsByAddressCount(ctx Context, address string) (uint64, error) {
...
+		return nil, nil
+	}
+
 	db, err := ctx.db()
 	if err != nil {
 		return nil, err
 	}
+	var transactions []*dbmodels.Transaction
+	query := db.Model(&transactions).
+		ColumnExpr("DISTINCT(transaction.id)").
```
Comment (someone235) [inline dbaccess/transaction.go]: Why sometimes you do `err = query.Select("DISTINCT transactions.*")` and sometimes you do `ColumnExpr("DISTINCT(transaction.id)")`? Can you explain the difference?

File: dbaccess/transaction.go [hunk 11]
```diff
@@ -268,25 +266,23 @@ func UpdateTransactionAcceptingBlockID(ctx Context, transactionID uint64, accept
...
 
-func joinTxInputsTxOutputsAndAddresses(query *gorm.DB) *gorm.DB {
+func joinTxInputsTxOutputsAndAddresses(query *orm.Query) *orm.Query {
 	return query.
-		Joins("LEFT JOIN `transaction_outputs` ON `transaction_outputs`.`transaction_id` = `transactions`.`id`").
-		Joins("LEFT JOIN `addresses` AS `out_addresses` ON `out_addresses`.`id` = `transaction_outputs`.`address_id`").
-		Joins("LEFT JOIN `transaction_inputs` ON `transaction_inputs`.`transaction_id` = `transactions`.`id`").
-		Joins("LEFT JOIN `transaction_outputs` AS `inputs_outs` ON `inputs_outs`.`id` = `transaction_inputs`.`previous_transaction_output_id`").
-		Joins("LEFT JOIN `addresses` AS `in_addresses` ON `in_addresses`.`id` = `inputs_outs`.`address_id`")
+		Join("LEFT JOIN transaction_outputs ON transaction_outputs.transaction_id = transaction.id").
```
Comment (someone235) [inline dbaccess/transaction.go]: Why in some places you use JoinOn and here you don't?

File: dbaccess/transaction.go [hunk 12]
```diff
@@ -268,25 +266,23 @@ func UpdateTransactionAcceptingBlockID(ctx Context, transactionID uint64, accept
...
+		Join("LEFT JOIN addresses AS out_addresses ON out_addresses.id = transaction_outputs.address_id").
+		Join("LEFT JOIN transaction_inputs ON transaction_inputs.transaction_id = transaction.id").
```
Comment (someone235) [inline dbaccess/transaction.go]: shouldn't it be transactions.id and not transaction.id?

File: dbaccess/transaction.go (hunk 11 above)
Comment (someone235) [inline dbaccess/transaction.go]: shouldn't it be transactions.id and not transaction.id?

File: dbaccess/transaction_output.go [hunk 13]
```diff
@@ -52,16 +51,15 @@ func TransactionOutputsByOutpoints(ctx Context, outpoints []*Outpoint) ([]*dbmod
...
-			Find(&dbPreviousTransactionsOutputsChunk)
-		dbErrors := dbResult.GetErrors()
-
-		if httpserverutils.HasDBError(dbErrors) {
-			return nil, httpserverutils.NewErrorFromDBErrors("failed to find previous transaction outputs: ", dbErrors)
+		err = db.Model(&dbPreviousTransactionsOutputsChunk).
+			Join("LEFT JOIN transactions").
+			JoinOn("transactions.id = transaction_output.transaction_id").
+			Where("(transactions.transaction_id, transaction_output.index) in (?)", pg.In(chunk)).
+			Relation("Transaction").
```
Comment (someone235) [inline dbaccess/transaction_output.go]: Use a constant

File: dbmodels/dbmodels.go [hunk 14]
```diff
@@ -78,28 +78,28 @@ var RawBlockFieldNames = struct {
...
+	LockTime           []byte
 	SubnetworkID       uint64
 	Subnetwork         Subnetwork
-	Gas                uint64
-	PayloadHash        string
-	Payload            []byte
+	Gas                uint64 `pg:",use_zero"`
+	PayloadHash        string `pg:",use_zero"`
+	Payload            []byte `pg:",use_zero"`
 	Mass               uint64
```
Comment (someone235) [inline dbmodels/dbmodels.go]: Maybe we should use use_zero here? I'm not sure. I would expect all non-ID integers to use the use_zero flag

File: go.mod [hunk 15]
```diff
@@ -4,14 +4,14 @@ go 1.13
 
 require (
 	github.com/eclipse/paho.mqtt.golang v1.2.0
+	github.com/go-pg/pg/v9 v9.1.3
 	github.com/golang-migrate/migrate/v4 v4.7.1
 	github.com/gorilla/handlers v1.4.2
 	github.com/gorilla/mux v1.7.3
 	github.com/jessevdk/go-flags v1.4.0
 	github.com/jinzhu/gorm v1.9.11
```
Comment (someone235) [inline go.mod]: get rid of gorm

File: kasparovd/controllers/block.go [hunk 16]
```diff
@@ -41,7 +41,7 @@ func GetBlockByHashHandler(blockHash string) (interface{}, error) {
 }
 
 // GetBlocksHandler searches for all blocks
-func GetBlocksHandler(orderString string, skip uint64, limit uint64) (interface{}, error) {
+func GetBlocksHandler(orderString string, skip int, limit int) (interface{}, error) {
```
Comment (someone235) [inline kasparovd/controllers/block.go]: shouldn't it be still uint?

File: kasparovsyncd/sync/sync.go [hunk 17]
```diff
@@ -188,11 +188,10 @@ func updateSelectedParentChain(client *jsonrpc.Client, removedChainHashes []stri
 	if err != nil {
 		return err
 	}
-	defer dbTx.RollbackUnlessCommitted()
-
 	for _, removedHash := range removedChainHashes {
 		err := updateRemovedChainHashes(dbTx, removedHash)
 		if err != nil {
+			dbTx.Rollback()
```
Comment (someone235) [inline kasparovsyncd/sync/sync.go]: Please use defer here and in all similar places

**Review by someone235 — CHANGES_REQUESTED** [review]
Please check that all API endpoints and MQTT topics work without errors

File: config/config.go [hunk 18]
```diff
@@ -11,15 +11,19 @@ import (
...
 )
 
 // KasparovFlags holds configuration common to both the Kasparov server and the Kasparov daemon.
 type KasparovFlags struct {
 	ShowVersion bool   `short:"V" long:"version" description:"Display version information and exit"`
 	LogDir      string `long:"logdir" description:"Directory to log output."`
 	DebugLevel  string `short:"d" long:"debuglevel" description:"Set log level {trace, debug, info, warn, error, critical}"`
-	DBAddress   string `long:"dbaddress" description:"Database address"`
+	DBHost      string `long:"dbhost" description:"Database host"`
+	DBPort      string `long:"dbport" description:"Database port"`
```
Comment (svarogg) [inline config/config.go]: Please mention the defaults in the description

File: database/database.go [hunk 19]
```diff
@@ -3,44 +3,33 @@ package database
...
-type gormLogger struct{}
-
-func (l gormLogger) Print(v ...interface{}) {
-	str := fmt.Sprint(v...)
-	log.Errorf(str)
-}
-
-// Connect connects to the database mentioned in
-// config variable.
+// Connect connects to the database mentioned in config variable.
```
Comment (svarogg) [inline database/database.go]: Grammar: "mentioned in **the** config variable"
Another option: "montioned in cfg"

File: config/config.go (hunk 1 above)
Comment (dereeno) [reply config/config.go]: It is used in `buildMigratorConnectionString` database.go

File: database/database.go (hunk 2 above)
Comment (dereeno) [reply database/database.go]: Missed that, thanks

File: dbaccess/transaction.go (hunk 12 above)
Comment (dereeno) [reply dbaccess/transaction.go]: No, go-pg uses singular form for the main table

File: dbaccess/transaction.go (hunk 10 above)
Comment (dereeno) [reply dbaccess/transaction.go]: This was a mistake `ColumnExpr` is the right way to go

File: dbaccess/block.go (hunk 7 above)
Comment (dereeno) [reply dbaccess/block.go:117]: No, MAX queries returns a number (maximum value of a column)

File: dbaccess/block.go (hunk 4 above)
Comment (dereeno) [reply dbaccess/block.go]: agree

File: database/migrations/000001_create_blocks_table.up.sql (hunk 3 above)
Comment (dereeno) [reply database/migrations/000001_create_blocks_table.up.sql]: I'm not sure postgres can do that, tried looking for that

File: dbmodels/dbmodels.go (hunk 14 above)
Comment (dereeno) [reply dbmodels/dbmodels.go]: This is for fields with `NOT NULL` constraint. I don't think we would want to allow some fields to be defaulted to zero.
I'll add to the relevant fields. (Mass is one of them)

File: database/dbtx.go [hunk 20]
```diff
@@ -1,47 +1,54 @@
...
 
 import (
 	"github.com/go-pg/pg/v9"
-	"github.com/kaspanet/kasparov/database"
+	"github.com/go-pg/pg/v9/orm"
 )
 
+// DbTx is an interface type implemented both by pg.DB and pg.Tx.
+// We use DbTX to execute db operations with or without transaction context.
+type DbTx interface {
```
Comment (someone235) [inline database/dbtx.go]: call it DBTx or DatabaseTx or DatabaseTransaction

File: database/dbtx.go [hunk 21]
```diff
@@ -1,47 +1,54 @@
...
+	Update(model interface{}) error
+	Delete(model interface{}) error
+}
+
 // Context is an interface type representing the context in which queries run, currently relating to the
 // existence or non-existence of a database transaction
 // Call `.NoTx()` or `.NewTx()` to acquire a Context
 type Context interface {
-	db() (*pg.DB, error)
+	Db() (DbTx, error)
```
Comment (someone235) [inline database/dbtx.go]: Db -> DB

File: database/dbtx.go [hunk 22]
```diff
@@ -1,47 +1,54 @@
...
+// Db returns a db instance
+func (*noTxContext) Db() (DbTx, error) {
+	return DB()
 }
 
 // TxContext represents a database context with an attached database transaction
-type TxContext struct{ dbInstance *pg.DB }
+type TxContext struct {
+	tx       *pg.Tx
+	Commited bool
```
Comment (someone235) [inline database/dbtx.go]: Commited -> Committed

Comment (someone235) [reply database/dbtx.go]: And please unexport it and instead export RollbackUnlessCommitted

File: dbmodels/dbmodels.go [hunk 23]
```diff
@@ -96,7 +96,7 @@ type Transaction struct {
 	Gas                uint64 `pg:",use_zero"`
 	PayloadHash        string `pg:",use_zero"`
 	Payload            []byte `pg:",use_zero"`
-	Mass               uint64
+	Mass               uint64 `pg:",use_zero"`
```
Comment (someone235) [inline dbmodels/dbmodels.go:99]: So eventually we don't want it to use use_zero here, right?

File: config/config.go (hunk 1 above)
Comment (someone235) [reply config/config.go]: You should also use it for `pg.Connect(&pg.Options{`

File: database/migrations/000001_create_blocks_table.up.sql (hunk 3 above)
Comment (someone235) [reply database/migrations/000001_create_blocks_table.up.sql]: Just put maxuint32 as a number

File: dbaccess/transaction_output.go (hunk 13 above)
Comment (someone235) [reply dbaccess/transaction_output.go]: It's not fixed

File: database/dbtx.go [hunk 24]
```diff
@@ -0,0 +1,81 @@
+package database
+
+import (
+	"github.com/go-pg/pg/v9"
+	"github.com/go-pg/pg/v9/orm"
+)
+
+// DBTx is an interface type implemented both by pg.DB and pg.Tx.
+// We use DBTx to execute db operations with or without transaction context.
+type DBTx interface {
```
Comment (svarogg) [inline database/dbtx.go]: This interface is not necessarily implemented by TX, it's also implemented by just DB.
Please find a more suitable name

File: database/dbtx.go [hunk 25]
```diff
@@ -0,0 +1,81 @@
...
+	ctx.committed = true
+	return ctx.tx.Commit()
+}
+
+// Rollback rolls-back the transaction attached to this TxContext
+func (ctx *TxContext) Rollback() error {
+	return ctx.tx.Rollback()
+}
+
+func (ctx *TxContext) RollbackUnlessCommitted() error {
```
Comment (svarogg) [inline database/dbtx.go]: Please add comment to RollbackUnlessCommitted

File: database/dbtx.go [hunk 26]
```diff
@@ -0,0 +1,81 @@
...
+
+// Commit commits the transaction attached to this TxContext
+func (ctx *TxContext) Commit() error {
+	ctx.committed = true
+	return ctx.tx.Commit()
+}
+
+// Rollback rolls-back the transaction attached to this TxContext
+func (ctx *TxContext) Rollback() error {
+	return ctx.tx.Rollback()
```
Comment (svarogg) [inline database/dbtx.go]: Please return error if `ctx.committed == true`

File: dbaccess/block.go [hunk 27]
```diff
@@ -2,156 +2,144 @@ package dbaccess
...
 	if err != nil {
 		return nil, err
 	}
 
-	query := db.
-		Offset(skip).
-		Limit(limit)
+	var blocks []*dbmodels.Block
+	query := db.Model(&blocks).
+		Offset(int(skip)).
```
Comment (svarogg) [inline dbaccess/block.go:66]: This (very theoretically) can be a problem.
In theory we allow 2^64 blocks in a dag. So in theory 1 day int might not be enough.

It's probably not a problem, but let's give it a short discussion in morning meeting

Comment (svarogg) [reply dbaccess/block.go:66]: However, please see what happens if you pass negative values here(which will happen if user supplies numbers above 2^32), I guess nothing terrible, but please make sure it is.

File: dbaccess/block.go [hunk 28]
```diff
@@ -2,156 +2,144 @@ package dbaccess
...
 
-	blocks := []*dbmodels.Block{}
-	dbResult := query.Find(&blocks)
-
-	dbErrors := dbResult.GetErrors()
-	if httpserverutils.HasDBError(dbErrors) {
-		return nil, httpserverutils.NewErrorFromDBErrors("some errors were encountered when loading blocks from the database:",
-			dbResult.GetErrors())
+	err = query.Select()
+	if err != nil {
```
Comment (svarogg) [inline dbaccess/block.go:76]: I think you need to make sure err is not pg.ErrNoRows here, and `return nil, nil` in these cases. Here and anywhere else relevant.

File: dbaccess/block.go [hunk 29]
```diff
@@ -160,99 +148,92 @@ func BluestBlock(ctx Context) (*dbmodels.Block, error) {
 // UpdateBlocksAcceptedByAcceptingBlock updates all blocks which are currently accepted by `currentAcceptingBlockID`
 // to be accepted by `newAcceptingBlockID`.
 // `newAcceptingBlockID` can be set nil.
-func UpdateBlocksAcceptedByAcceptingBlock(ctx Context, currentAcceptingBlockID uint64, newAcceptingBlockID *uint64) error {
-	db, err := ctx.db()
+func UpdateBlocksAcceptedByAcceptingBlock(ctx database.Context, currentAcceptingBlockID uint64, newAcceptingBlockID *uint64) error {
```
Comment (svarogg) [inline dbaccess/block.go:151]: Does `.Save()` in go-pg work the same silly way as in GORM?

If not, maybe it's a good idea to convert all/some .Update functions to .Save functions, to minimize code-complexity.

File: kasparovsyncd/sync/sync.go [hunk 30]
```diff
@@ -412,21 +414,24 @@ func updateAddedChainBlocks(dbTx *dbaccess.TxContext, addedBlock *rpcmodel.Chain
...
 	if blockExists {
 		return nil
 	}
 
+	dbTx, err := database.NewTx()
+	if err != nil {
+		return err
+	}
+
+	defer func(dbTx *database.TxContext) {
```
Comment (svarogg) [inline kasparovsyncd/sync/sync.go]: Why do you need the closure here?
You can simply do `defer dbTx.RollbackUnlessCommited`, no?

**Review by svarogg — CHANGES_REQUESTED** [review]

File: kasparovsyncd/sync/sync.go (hunk 30 above)
Comment (dereeno) [reply kasparovsyncd/sync/sync.go]: I thought this is needed as we update dbtx, but since it's a pointer you're right. I'll fix that

File: database/migrations/000005_create_transactions_table.up.sql [hunk 31]
```diff
@@ -1,20 +1,21 @@
...
-    INDEX `idx_transactions_transaction_id` (`transaction_id`),
-    CONSTRAINT `fk_transactions_accepting_block_id`
-        FOREIGN KEY (`accepting_block_id`)
-            REFERENCES `blocks` (`id`)
+    id                 BIGSERIAL,
+    accepting_block_id BIGINT NULL,
+    transaction_hash   CHAR(64) NOT NULL,
+    transaction_id     CHAR(64) NOT NULL,
+    lock_time          BYTEA NOT NULL,
+    subnetwork_id      BIGINT CHECK (subnetwork_id >= 0) NOT NULL,
```
Comment (svarogg) [inline database/migrations/000005_create_transactions_table.up.sql]: I think we are missing foreign key for subnetwork_id.
And the moment you have the foreign key, it seems like the check constraint becomes redundant.

File: database/migrations/000006_create_transactions_to_blocks_table.up.sql [hunk 32]
```diff
@@ -1,14 +1,15 @@
...
-    INDEX `idx_transactions_to_blocks_index` (`index`),
-    CONSTRAINT `fk_transactions_to_blocks_block_id`
-        FOREIGN KEY (`block_id`)
-            REFERENCES `blocks` (`id`),
-    CONSTRAINT `fk_transactions_to_blocks_transaction_id`
-        FOREIGN KEY (`transaction_id`)
-            REFERENCES `transactions` (`id`)
+    transaction_id BIGINT NOT NULL,
+    block_id       BIGINT NOT NULL,
+    index BIGINT CHECK (index >= 0) NOT NULL,
```
Comment (svarogg) [inline database/migrations/000006_create_transactions_to_blocks_table.up.sql]: index is uint32, originally, so needs to make sure values stay within uint32 range

File: database/migrations/000008_create_transaction_outputs_table.up.sql [hunk 33]
```diff
@@ -1,18 +1,19 @@
...
-    INDEX `idx_transaction_outputs_transaction_id` (`transaction_id`),
-    CONSTRAINT `fk_transaction_outputs_transaction_id`
-        FOREIGN KEY (`transaction_id`)
-            REFERENCES `transactions` (`id`),
-    CONSTRAINT `fk_transaction_outputs_address_id`
-        FOREIGN KEY (`address_id`)
-            REFERENCES `addresses` (`id`)
+    id             BIGSERIAL,
+    transaction_id BIGINT NOT NULL,
+    index          BIGINT CHECK (index >= 0) NOT NULL,
```
Comment (svarogg) [inline database/migrations/000008_create_transaction_outputs_table.up.sql]: here too, index is uint32, originally, so needs to make sure values stay within uint32 range

File: database/migrations/000001_create_blocks_table.up.sql [hunk 34]
```diff
@@ -1,23 +1,24 @@
...
-    INDEX `idx_blocks_timestamp` (`timestamp`),
-    INDEX `idx_blocks_is_chain_block` (`is_chain_block`),
-    CONSTRAINT `fk_blocks_accepting_block_id`
-        FOREIGN KEY (`accepting_block_id`)
-            REFERENCES `blocks` (`id`)
-);
\ No newline at end of file
+    id                      BIGSERIAL,
+    block_hash              CHAR(64) NOT NULL,
+    accepting_block_id      BIGINT CHECK (accepting_block_id > 0) NULL,
```
Comment (svarogg) [inline database/migrations/000001_create_blocks_table.up.sql]: You don't need check contraint on fields that are foreign-keyed to BIGSERIALs

File: database/dbtx.go [hunk 35]
```diff
@@ -0,0 +1,86 @@
+package database
```
Comment (svarogg) [inline database/dbtx.go]: Please also rename the file dbtx.go to something else.
I think the most fitting would be context.go, as it 99% deals with contexts.

File: dbaccess/block.go (hunk 27 above)
Comment (svarogg) [reply dbaccess/block.go:66]: It seems like you missed/forgot about this comment

File: dbaccess/block.go (hunk 29 above)
Comment (svarogg) [reply dbaccess/block.go:151]: You also didn't answer this one.

**Review by svarogg — CHANGES_REQUESTED** [review]

File: dbaccess/block.go (hunk 27 above)
Comment (dereeno) [reply dbaccess/block.go:66]: In that case we return:
```
{"errorCode":422,"errorMessage":"Couldn't parse the 'skip' query parameter"}
```

File: dbaccess/block.go (hunk 28 above)
Comment (dereeno) [reply dbaccess/block.go:76]: Done

File: dbaccess/block.go (hunk 29 above)
Comment (dereeno) [reply dbaccess/block.go:151]: I wrote on Telegram that I think this one should be done in a separate ticket. I can try to squeeze it tomorrow, but let's first merge this one, and make sure nothing is broken before doing any refactors.

File: database/migrations/000005_create_transactions_table.up.sql (hunk 31 above)
Comment (dereeno) [reply database/migrations/000005_create_transactions_table.up.sql]: Fixed, thanks

**Review by svarogg — DISMISSED** [review]

File: dbaccess/block.go [hunk 36]
```diff
@@ -71,8 +71,11 @@ func Blocks(ctx database.Context, order Order, skip uint64, limit uint64, preloa
 	}
 
 	query = preloadFields(query, preloadedFields)
-
 	err = query.Select()
+	if err == pg.ErrNoRows {
```
Comment (someone235) [inline dbaccess/block.go]: I thought we said it shouldn't return ErrNoRows in here?

File: dbmodels/dbmodels.go (hunk 23 above)
Comment (someone235) [reply dbmodels/dbmodels.go:99]: It's still here

**Review by someone235 — CHANGES_REQUESTED** [review]

File: config/config.go [hunk 37]
```diff
@@ -11,15 +11,19 @@ import (
 
 var (
 	// Default configuration options
-	defaultDBAddress = "localhost:3306"
+	defaultDBHost    = "localhost"
```
Comment (someone235) [inline config/config.go]: Why it's defined twice?

**Review by someone235 — CHANGES_REQUESTED** [review]

**Review by someone235 — DISMISSED** [review]

**Review by someone235 — DISMISSED** [review]

**Review by someone235 — DISMISSED** [review]

File: dbmodels/dbmodels.go (hunk 23 above)
Comment (svarogg) [reply dbmodels/dbmodels.go:99]: @someone235 We decided with Dan it's better to leave it there. Lets fight face-to-face if you disagree.

**Review by svarogg — DISMISSED** [review]

File: dbmodels/dbmodels.go [hunk 38]
```diff
@@ -11,16 +11,16 @@ type FieldName string
...
 	AcceptingBlock       *Block
-	Version              int32
-	HashMerkleRoot       string
-	AcceptedIDMerkleRoot string
-	UTXOCommitment       string
-	Timestamp            time.Time
-	Bits                 uint32
-	Nonce                []byte
+	Version              int32          `pg:",use_zero"`
+	HashMerkleRoot       string         `pg:",use_zero"`
```
Comment (svarogg) [inline dbmodels/dbmodels.go:18]: Do we need use_zero in stringy, etc fields?

**Review by svarogg — CHANGES_REQUESTED** [review]

**Review by svarogg — DISMISSED** [review]

**Review by someone235 — APPROVED** [review]

**Review by svarogg — APPROVED** [review]
